### PR TITLE
add no-constant-condition rule

### PR DESCRIPTION
**Pull request by humancalico**

### CODE REVIEW

File: src/rules/no_constant_condition.rs [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+      _ => false,
+    }
+  }
+
+  fn report(&self, condition: &Expr) {
+    if self.is_constant(condition, None, true) {
+      let span = condition.span();
+      self.add_diagnostic(span);
+    } else {
+    }
```
Comment (bartlomieju) [inline src/rules/no_constant_condition.rs:164]: nitpick: remove `else` (should have been picked up by clippy)

File: src/rules/no_constant_condition.rs [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+        (in_boolean_position
+          && tpl.quasis.iter().any(|quasi| match &quasi.cooked {
+            Some(str) => !str.is_empty(),
+            None => false,
+          }))
+          || tpl.exprs.iter().all(|expr| {
+            self.is_constant(expr, parent_node, in_boolean_position)
+          })
+      }
+      // TODO(humancalico) confirm in_boolean_position here
```
Comment (bartlomieju) [inline src/rules/no_constant_condition.rs:94]: is this TODO still relevant? What do we need to confirm here?

File: src/rules/no_constant_condition.rs [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+    &mut self,
+    if_stmt: &swc_ecma_ast::IfStmt,
+    _parent: &dyn Node,
+  ) {
+    self.report(&if_stmt.test)
+  }
+  fn visit_while_stmt(
+    &mut self,
+    while_stmt: &swc_ecma_ast::WhileStmt,
+    __parent: &dyn Node,
```
Comment (bartlomieju) [inline src/rules/no_constant_condition.rs:186]: Two nitpicks:
1. use single underscore
2. add empty line between the methods

File: src/rules/no_constant_condition.rs [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+    assert_lint_err::<NoConstantCondition>(r#"for(;`foo${bar}`;);"#, 5);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(true)"#, 10);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(t = -2)"#, 10);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(``)"#, 10);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(`foo`)"#, 10);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(`foo${bar}`)"#, 10);
+
+    assert_lint_err::<NoConstantCondition>(r#"true ? 1 : 2;"#, 0);
+    // FIXME(humancalico) Is it supposed to be on column 4
+    assert_lint_err::<NoConstantCondition>(r#"q = 0 ? 1 : 2;"#, 4);
```
Comment (bartlomieju) [inline src/rules/no_constant_condition.rs:344]: Which column should it be on?

File: src/rules/no_constant_condition.rs [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo(){for(;true;){yield 'foo';}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo(){do{yield 'foo';}while(true)}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo(){while (true) { while(true) {yield;}}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() {for (; yield; ) {}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() {for (; ; yield) {}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() {while (true) {function* foo() {yield;}yield;}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() { for (let x = yield; x < 10; x++) {yield;}yield;}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() { for (let x = yield; ; x++) { yield; }}"#);
+
+    assert_lint_err::<NoConstantCondition>(r#"for(;true;);"#, 5);
```
Comment (bartlomieju) [inline src/rules/no_constant_condition.rs:332]: Nitpick: put failing test cases in a separate test function

File: src/rules/no_constant_condition.rs [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+
+    // string literals
+    assert_lint_err::<NoConstantCondition>(r#"if('str1' || 'str2'){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if('str1' && 'str2'){}"#, 3);
+    // TODO(humancalico) make these test pass
+    // assert_lint_err::<NoConstantCondition>(r#"if(abc==='str' || 'str'){}"#, 3);
+    // assert_lint_err::<NoConstantCondition>(r#"if(a || 'str'){}"#, 3);
+
+    // assert_lint_err::<NoConstantCondition>(r#"function* foo(){while(true){} yield 'foo';}"#, );
+    // assert_lint_err::<NoConstantCondition>(r#"function* foo(){while(true){if (true) {yield 'foo';}}}"#, );
```
Comment (bartlomieju) [inline src/rules/no_constant_condition.rs:421]: What's the problem with these tests?

File: src/rules/no_constant_condition.rs (hunk 2 above)
Comment (humancalico) [reply src/rules/no_constant_condition.rs:94]: There is no Expression for `Paren` in the original `eslint` implementation and I was not sure what in_boolean_position does (it says `false` if checking branch of a condition. `true` in all other cases) so I am not sure if it'll be `false` here.

File: src/rules/no_constant_condition.rs (hunk 4 above)
Comment (humancalico) [reply src/rules/no_constant_condition.rs:344]: Just checked it on https://eslint.org/demo and it to be correct. I'll remove the FIXME tag.

File: src/rules/no_constant_condition.rs (hunk 6 above)
Comment (humancalico) [reply src/rules/no_constant_condition.rs:421]: These should work once [this](https://github.com/eslint/eslint/blob/f4d7b9e1a599346b2f21ff9de003b311b51411e6/lib/rules/no-constant-condition.js#L135-L146) condition is added

**Comment by humancalico** [comment]
Done @bartlomieju

**Comment by bartlomieju** [comment]
LGTM, thanks @humancalico!

**Pull request merged**